### ISSUE TITLE
Issue with manager.startDeviceScan callback not being fired on iOS when used outside manager.onStateChange callback

### ISSUE THREAD

**Issue by qhdcrobey** [Polidea/react-native-ble-plx#773]
## Expected Behaviour

I have created a button that sets a state variable, `scanning` to `true`. I am using an effect that has this state variable as a dependancy and will start scanning when it is set to `true`:

```
  useEffect(() => {
    if(scanning) {
      setDevices([]);
      scanAndConnect();
    }
  }, [scanning]);
```

My scanAndConnect() function:

```
const scanAndConnect = function() {
    manager.startDeviceScan(null, null, (error, device) => {
      if (error) {
        // Handle error (scanning will be stopped automatically)
        return;
      }

      console.log(device);
    });
  };
```

However I am finding that this is not working (the function gets called however it seems like the listener callback is never fired), and the only way to get the scanning working (on iOS) is to fire my scanAndConnect() function inside the `manager.onStateChange` callback:

```
  const subscription = manager.onStateChange((state) => {
    if (state === 'PoweredOn' && scanning === false) {
      setCanScan(true);
      subscription.remove();
      scanAndConnect(); #the function properly gets called here
    }
  }, true);
```

Please describe the behaviour you are expecting

## Current Behaviour

See above

### Steps to Reproduce

Please provide detailed steps for reproducing the issue.

See above

### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions. Both JS and platform logs can be enabled via [setLogLevel](https://polidea.github.io/react-native-ble-plx/#blemanagersetloglevel) function call. 

* Library version: ^2.0.1
* Platform: Android/iOS.
* Platform logs (logcat/XCode): 

These are some errors that I believe are related:

```
error	16:28:33.457562+1000	Boilerplate	API MISUSE: <CBCentralManager: 0x2839cc800> has no restore identifier but the delegate implements the centralManager:willRestoreState: method. Restoring will not be supported
error	16:28:33.486466+1000	Boilerplate	API MISUSE: <CBCentralManager: 0x2839edb00> has no restore identifier but the delegate implements the centralManager:willRestoreState: method. Restoring will not be supported
error	16:28:33.489156+1000	Boilerplate	XPC connection invalid
error	16:28:33.503813+1000	Boilerplate	API MISUSE: <CBCentralManager: 0x2839c8980> has no restore identifier but the delegate implements the centralManager:willRestoreState: method. Restoring will not be supported
error	16:28:33.506644+1000	Boilerplate	XPC connection invalid

error	16:29:29.509341+1000	Boilerplate	API MISUSE: <CBCentralManager: 0x2839e6280> has no restore identifier but the delegate implements the centralManager:willRestoreState: method. Restoring will not be supported
error	16:29:29.522149+1000	Boilerplate	XPC connection invalid
error	16:29:29.573470+1000	Boilerplate	API MISUSE: <CBCentralManager: 0x2839c9c80> has no restore identifier but the delegate implements the centralManager:willRestoreState: method. Restoring will not be supported
error	16:29:29.585231+1000	Boilerplate	XPC connection invalid

```

* JS logs: No obvious errors or anything to point to an issue
* Contents of the `package.json` file: 

```
{
  "name": "Boilerplate",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "react-native start",
    "test": "jest",
    "cocoapods": "cd ios && pod install && cd ..",
    "ios": "react-native run-ios --simulator=\"iPhone 8\"",
    "build:ios": "react-native bundle --entry-file='index.js' --bundle-output='./ios/main.jsbundle' --dev=false --platform='ios'",
    "android": "react-native run-android",
    "rename": "react-native-rename",
    "redux:remove": "yarn remove redux react-redux redux-thunk",
    "typescript:add": "sh scripts/add-typescript/script.sh",
    "lint": "eslint ."
  },
  "dependencies": {
    "@react-native-async-storage/async-storage": "^1.13.2",
    "@react-native-community/masked-view": "^0.1.10",
    "@react-navigation/drawer": "^5.10.4",
    "@react-navigation/native": "^5.2.4",
    "@react-navigation/stack": "^5.2.19",
    "appcenter": "3.1.2",
    "appcenter-analytics": "3.1.2",
    "appcenter-crashes": "3.1.2",
    "dayjs": "^1.9.5",
    "engine.io-client": "^4.0.4",
    "laravel-echo": "^1.9.0",
    "moment": "^2.29.1",
    "native-base": "^2.13.14",
    "react": "16.13.1",
    "react-native": "0.63.2",
    "react-native-asyncstorage": "^1.0.0",
    "react-native-ble-plx": "^2.0.1",
    "react-native-gesture-handler": "^1.8.0",
    "react-native-map-clustering": "^3.3.9",
    "react-native-maps": "^0.27.1",
    "react-native-paper": "^4.3.1",
    "react-native-permissions": "^3.0.0-beta.2",
    "react-native-reanimated": "^2.0.0-alpha.8",
    "react-native-safe-area-context": "^0.7.3",
    "react-native-screens": "^2.7.0",
    "react-native-swipe-gestures": "^1.0.5",
    "react-native-vector-icons": "^7.0.0",
    "react-redux": "^7.2.0",
    "redux": "^4.0.5",
    "redux-thunk": "^2.3.0",
    "socket.io-client": "^2.3.1",
    "util": "^0.12.3"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/runtime": "^7.8.4",
    "@react-native-community/eslint-config": "^1.1.0",
    "babel-jest": "^25.1.0",
    "babel-plugin-module-resolver": "^4.0.0",
    "eslint": "^6.5.1",
    "jest": "^25.1.0",
    "metro-react-native-babel-preset": "^0.59.0",
    "react-native-rename": "^2.4.1",
    "react-test-renderer": "16.13.1"
  },
  "jest": {
    "preset": "react-native"
  }
}
```

* Formatted code sample or link to a repository:

```js
import React, { useState, useEffect } from 'react';
import {Image, StyleSheet, TouchableHighlight, View} from 'react-native';
import { useSelector } from 'react-redux';
import Logo from '../../../assets/images/qhdc-logo-transparent.png';
import { MaterialColors, FontWeights, FontSizes } from '@app/theme';
import { BleManager } from 'react-native-ble-plx';
import { setAssetViewId, handleGetAllAssets } from '../../store/actions/index';
import { useDispatch } from 'react-redux';
import { getObject } from '../../utils/index';
import dayjs from 'dayjs';
import Sidebar from '../../components/Sidebar/index';
import { Avatar, Button, Card, Title, Paragraph, Text, Chip, ActivityIndicator, Colors } from 'react-native-paper';
import { useNavigation } from '@react-navigation/native';

var relativeTime = require('dayjs/plugin/relativeTime');
dayjs.extend(relativeTime);

function AssetScreen() {
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const { status, page, loginError, user, assets } = useSelector(({ data }) => data);
  const [devices, setDevices] = useState([]);
  const [scanning, setScanning] = useState(false);
  const [canScan, setCanScan] = useState(false);
  const [spidersPressed, setSpidersPressed] = useState(false);

  var assetMacs = {};

  const handleScan = () => {
    setScanning(true);
  };

  const manager = new BleManager();

  useEffect(() => {
    if(scanning) {
      setDevices([]);
      scanAndConnect();
    }
  }, [scanning]);

  const handleDeviceFound = async (device) => {
    delete device._manager;

    device.timestamp = dayjs().toISOString();
    setDevices(prevDevices => {
      let found = false;
      prevDevices.forEach((prevDevice, index) => {
        if (prevDevice.id === device.id) {
          found = true;
        }
      });
      const deviceMacNoColons = device.id.replace(/\:/g, '');

      if(typeof assetMacs[deviceMacNoColons] !== 'undefined') {
        device.name = assetMacs[deviceMacNoColons].name;
        device.device_id = assetMacs[deviceMacNoColons].device_id;
        device.asset_id = assetMacs[deviceMacNoColons].id;
        device.type = assetMacs[deviceMacNoColons].type;
        device.site = assetMacs[deviceMacNoColons].site;
      }

      if (found) {
        return [
          ...prevDevices
        ];
      } else {
        return [
          ...prevDevices,
          device
        ];
      }
    });
  };

  const scanAndConnect = function() {
    manager.startDeviceScan(null, null, (error, device) => {
      if (error) {
        // Handle error (scanning will be stopped automatically)
        return;
      }

      console.log(device);
    });
  };

  const subscription = manager.onStateChange((state) => {
    if (state === 'PoweredOn' && scanning === false) {
      setCanScan(true);
      subscription.remove();
    }
  }, true);

  function ScanButton() {
    if (scanning) {
      return (
        <Button style={{width:200}} mode="contained" loading={true} onPress={() => setScanning(false)}>
          Scanning...
        </Button>
      );
    }
    return (
      <Button style={{width:200}} mode="contained" onPress={() => handleScan()}>
        Scan
      </Button>
    );
  }

  const handleDevicePressed = (deviceId) => {
    dispatch(setAssetViewId(deviceId));
    navigation.navigate('AssetViewScreen');
  };

  return (
    <View style={{zIndex: 2, paddingLeft: 20, backgroundColor: '#fff', position: 'relative', height:'100%'}}>
      <View style={{zIndex: 2, position: 'relative', flex: 1, flexDirection:'row'}}>
        <View style={{position: 'relative', zIndex:2, paddingTop:20, backgroundColor:'#fff', height:'100%', flex:1}}>
          { canScan &&
          <View style={{marginTop: 5}}>
            { ScanButton() }
          </View>
          }

          {
            spidersPressed &&
            <View>
              <Text>
                { JSON.stringify(assets) }
              </Text>
            </View>
          }

          <View>
            <View style={{flexDirection: 'column', alignItems: 'flex-start', justifyContent: 'flex-start'}}>
              { devices.sort((a, b) => {
                if (a.rssi > b.rssi) {
                  return -1;
                } else if(a.rssi < b.rssi) {
                  return 1;
                }
                return 0;
              }).map(function(device) {
                if(typeof device.device_id !== 'undefined') {
                  return (
                    <Chip key={device.id} style={{marginTop:5}} icon="information" onPress={() => handleDevicePressed(device.asset_id)}>
                      {device.name + ' - ' + device.type.name} - seen { dayjs(device.timestamp).toNow(true) } ago @ {device.rssi}
                    </Chip>
                  );
                }
              }) }
            </View>
          </View>
        </View>
      </View>
    </View>
  );
}

export default AssetScreen;
```

**Comment by dariuszseweryn** [comment]
Let me cite the doc of `BleManager`

> /**
>  *
>  * BleManager is an entry point for react-native-ble-plx library. It provides all means to discover and work with
>  * {@link Device} instances. It should be initialized only once with `new` keyword and method
>  * {@link #blemanagerdestroy|destroy()} should be called on its instance when user wants to deallocate all resources.
>  *
>  * In case you want to properly support Background Mode, you should provide `restoreStateIdentifier` and
>  * `restoreStateFunction` in {@link BleManagerOptions}.
>  *
>  * @example
>  * const manager = new BleManager();
>  * // ... work with BLE manager ...
>  * manager.destroy();
>  */

You are using RN hooks. Guess what may be happening in your code and what can be done to prevent it?

**Comment by qhdcrobey** [comment]
@dariuszseweryn Thanks for your comment... it did lead me to discover this issue https://github.com/Polidea/react-native-ble-plx/issues/603 where it looks like there are issues with the way I am doing things.. seems like my useEffect may be getting called too early before the manager is ready, also looks like I need to pass the `restoreStateIdentifier` and `restoreStateFunction`... are you able to point me in the right direction?

**Comment by dariuszseweryn** [comment]
> It should be initialized only once with new keyword and method
{@link #blemanagerdestroy|destroy()} should be called on its instance when user wants to deallocate all resources

You are creating a new manager with each render in the above mentioned code. Waiting for the correct state is also important.

**Comment by qhdcrobey** [comment]
@dariuszseweryn Thank you again. I will close this issue until I implement the changes that you have nudged me towards making. Really appreciate the help. If I am still having troubles I will reopen it / create a new issue.
